src/arena.c

```diff
@@ -1369,63 +1319,108 @@ arena_bin_choose_lock(tsdn_t *tsdn, arena_t *arena, szind_t binind,
-			   empty_position - nfill + i);
-		} else {
-			cnt = 1;
-			void *ptr = arena_bin_malloc_hard(tsdn, arena, bin,
-			    binind, binshard);
+
+	/* Whether tried allocating in the _current_ refill iteration. */
+	bool cur_iter_alloc = false;
+	edata_t *fresh_slab = NULL;
+	unsigned filled = 0;
+	do {
+		unsigned cnt, binshard;
+		bin_t *bin = arena_bin_choose_lock(tsdn, arena, binind,
+		    &binshard);
+		bool terminate = true;
+		for (; filled < nfill; filled += cnt,
+		    cur_iter_alloc = false) {
+			edata_t *slabcur = bin->slabcur;
+			if (slabcur != NULL && edata_nfree_get(slabcur) > 0) {
+				unsigned tofill = nfill - filled;
+				unsigned nfree = edata_nfree_get(slabcur);
+				cnt = tofill < nfree ? tofill : nfree;
+				arena_slab_reg_alloc_batch(slabcur, bin_info,
+				    cnt, empty_position - tofill);
+				continue;
+			}
+			/* Try refill slabcur from nonfull first. */
+			cnt = 0;
+			if (!arena_bin_refill_slabcur_no_fresh_slab(tsdn, arena,
+			    bin)) {
+				assert(bin->slabcur != NULL);
+				continue;
+			}
+
+			/* See if a new slab was allocated already. */
+			if (fresh_slab != NULL) {
+				arena_bin_refill_slabcur_with_fresh_slab(tsdn,
+				    arena, bin, binind, fresh_slab);
+				assert(bin->slabcur != NULL);
+				fresh_slab = NULL;
+				continue;
+			}
+
+			/* ... and then try getting a new slab. */
+			if (!cur_iter_alloc) {
+				/* Alloc a new slab then come back. */
+				assert(bin->slabcur == NULL);
+				assert(fresh_slab == NULL);
+				terminate = false;
+				break;
+			}
+
+			assert(cur_iter_alloc && fresh_slab == NULL);
 			/*
 			 * OOM.  tbin->avail isn't yet filled down to its first
 			 * element, so the successful allocations (if any) must
 			 * be moved just before tbin->avail before bailing out.
 			 */
-			if (ptr == NULL) {
-				if (i > 0) {
-					memmove(empty_position - i,
-						empty_position - nfill,
-						i * sizeof(void *));
-				}
-				break;
+			if (filled > 0) {
+				memmove(empty_position - filled,
+				    empty_position - nfill,
+				    filled * sizeof(void *));
 			}
-			/* Insert such that low regions get used first. */
-			*(empty_position - nfill + i) = ptr;
+			assert(terminate);
+			break;
+		} /* for (filled < nfill) loop. */
+
+		if (terminate && config_stats) {
+			bin->stats.nmalloc += filled;
+			bin->stats.nrequests += tbin->tstats.nrequests;
+			bin->stats.curregs += filled;
+			bin->stats.nfills++;
+			tbin->tstats.nrequests = 0;
 		}
-		if (config_fill && unlikely(opt_junk_alloc)) {
-			for (unsigned j = 0; j < cnt; j++) {
-				void* ptr = *(empty_position - nfill + i + j);
-				arena_alloc_junk_small(ptr, &bin_infos[binind],
-							true);
-			}
+		malloc_mutex_unlock(tsdn, &bin->lock);
+
+		if (terminate) {
+			break;
+		}
+		assert(!cur_iter_alloc && fresh_slab == NULL);
+		fresh_slab = arena_slab_alloc(tsdn, arena, binind, binshard,
+		    bin_info);
+		cur_iter_alloc = true;
+		/* fresh_slab NULL case handled in the for loop. */
+	} while (true);
```

Comment:
This two-level loop is one of the most mysterious logic I've ever read! :) I think in any case it would be helpful to add some more comments.

On a high level, I can see that you're getting rid of the `arena_bin_malloc_hard()` call in the middle, and to fulfill its purpose you're calling `arena_slab_alloc()` at the end, outside of the bin lock.

I'm not quite sure if I'm understanding your usage of the two booleans correctly. Let me confirm:

`terminate`: on a higher level, I think `terminate = true` if either all `nfill` slots are fulfilled, or if there's no hope to believe that the next `arena_slab_alloc()` may succeed; otherwise `terminate = false`.

`cur_iter_alloc`: it's `true` only in the first iteration of the inner `for` loop of all non-first iterations of the outer `do`-`while` loop, and if it's `true`, you would lose hope and `terminate` whenever you cannot immediately find any available memory when under the bin lock.

I wonder: say if the `arena_slab_alloc()` call in a particular iteration of the `do`-`while` loop already OOMed, then in the next iteration of the `do`-`while` loop, how is it possible for the `for` loop to ever enter the second iteration? Probably only if it just happened that some other thread returned some available memory back during the time the bin lock is released but after `arena_slab_alloc()` returned `NULL`? If we ignore this possibility (if it's reasonably small), then we don't need this `cur_iter_alloc` at all, but rather just test for OOM on `arena_slab_alloc()` and break the `do`-`while` loop if it OOMed?

I understand that you're trying to keep the logic equivalent to what we had before: previously, `arena_bin_nonfull_slab_get()` calls `arena_bin_slabs_nonfull_tryget()` again even if `arena_slab_alloc()` OOMed, but at that time it was trivial to do so and it was pretty easy to understand. But now to do so, the logic becomes a puzzle to interpret. :) I think if we can get rid of the `cur_iter_alloc`, the logic can be much simpler (if the cost we're paying is negligible).

Reply:
Well, the possibility may not be negligible, but I guess, even if we double check after `arena_slab_alloc()` OOMed and it turns out that we still cannot get any available memory, it might be possible that some other thread may bring back some available memory before `arena_tcache_fill_small()` ends, so the double checking cannot provide us with any stronger guarantee.

Reply:
What you described is pretty much accurate. Indeed the `cur_iter_alloc` is to track, if the slab_alloc was attempted in this exact iteration of refill. You are totally right that this is the source of the complexity.

The refill can take quite some operations (think about refilling 100 items with mostly full slabs, which means dozens of slabs are used). One particular case the `cur_iter_alloc` solves is: 
1) current thread finds nothing in the bin, unlocks and tries slab_alloc;
2) slab_alloc returns nothing (I'll explain later why this still matters nowadays);
3) after relock, the thread is able to find available slabs in the bin, and keeps refilling;
4) after a while, the bin local slabs are all consumed, then the thread tries to see if `fresh_slab` has anything.

In this case, if we don't have `cur_iter_alloc` and conclude "tried allocating but didn't get fresh_slab; it must OOM", then it's based on things happened a while before (i.e. step 2). That's why I had the `cur_iter_alloc` to indicate if we just tried slab_alloc. Perhaps it should be named `recent_allocated` or `just_allocated`?

Then the benefit of doing is not quite obvious. If we consider only the OOM case on the powerful servers w/ overcommit, it doesn't matter at all. It comes down to two other scenarios:
1) if the current arena is hosting huge pages, in which case its limited resource and we should prioritize using it. Also it will be a contented resource (e.g. the idea David is trying, always prefer the dedicated arena w/ transparent huge pages). Here try harder before bailing out, makes sure the huge pages are tightly packed / utilized.
2) on small devices with limited memory and no overcommit, it can help avoid OOM.

Completely agree this needs better comments. However I'm still inclined to pay the complexity here because of the two cases above. WDYT?

Reply:
One thing forgot to mention was, arena_slab_alloc requires the `eset` lock, so it can potentially block and take a while, which is why retrying bin local helps.

Reply:
How about keeping the logic as is but trying a better way to illustrate the code logic? For example, how about the following strategy:
- Name variables *literally* stating their meaning,
- Define multiple variables each for a single meaning, and
- Put everything else into comments.

For example, one scheme I can think of is the following:
```
bool tried_arena_slab_alloc = false;
do {
    bool should_try_arena_slab_alloc = false;
    /* ... lock a bin ... */
    for (bool got_some_memory = false; filled < nfill; filled += cnt, got_some_memory = true) {
        /* ... try increasingly harder to find some available memory ... */
        /* ... and if we got nothing ... */
        if (!tried_arena_slab_alloc || got_some_memory) {
            /* ... explain ... */
            should_try_arena_slab_alloc = true;
            break;
        }
        /* ... handle OOM ... */
        break;
    }
    /* ... unlock the bin ... */
    if (!should_try_arena_slab_alloc) {
        break;
    }
    /* ... try arena_slab_alloc() ... */
    tried_arena_slab_alloc = true;
} while (true);
```
Essentially I'm using `should_try_arena_slab_alloc` for `!terminate`, and breaking `cur_iter_alloc` into the two separate components it was trying to capture, `tried_arena_slab_alloc` and `got_some_memory`. The specific wordings may not be the best, but I think it's along the right direction to make the logic more interpretable. Thoughts?

Reply:
Yeah let me try refactoring. One correction though, we may actually need slab_alloc multiple times. For example if a refill needs more items than a single slab holds; or if slab_alloc failed first but bin local resource became available (in which case, slab_alloc should be retried after exhausting bin resources, same as mention in my last comment). So the cur_iter_alloc actually means, was slab_alloc attempted very recently (will think about a better name as well).

Reply:
> So the cur_iter_alloc actually means, was slab_alloc attempted very recently

Right, but this "very recently" actually meant two possibilities:
- Directly at the end of the `do`-`while` loop;
- Indirectly afterwards.

Which was the reason why I thought it'd be easier to explicitly break it.

Reply:
I'm not following here. The "just_allocated" is true only if this current `for` iteration break out and does slab_alloc, and is false for the iterations afterwards. What do you mean by "Indirectly afterwards"?

Reply:
The problem with the code segment above is, `tried_arena_slab_alloc` is set to true and never reset, which means the slab_alloc can only be reached once.

Reply:
Discussed offline. I misunderstood the `got_some_memory` part. I think what Yinan suggested makes sense. Just updated.

Reply:
Now looks great!